### ISSUE TITLE
Custom yolov5 model Error

### ISSUE THREAD

**Issue by dingyan-dy** [marcoslucianops/DeepStream-Yolo#263]
I'm  facing the following problem (see the code below),  I tried with the custom pt file from my train with yolov5-5.0. I converted the .pt to .cgf and .wts by gen_wts_yoloV5.txt successly, however it seems that there is a problem somewhere else.
I mention that all requierements seems to be well installed; 
I'm running a Jetson Xavier NX with Jetpack 4.6.[L4T 32.6.1] and NVIDIA DeepStream SDK 6.0.0,
deepstream-app version 6.0.0
DeepStreamSDK 6.0.0
CUDA Driver Version: 10.2
CUDA Runtime Version: 10.2
TensorRT Version: 8.0
cuDNN Version: 8.2
libNVWarp360 Version: 2.0.1d3

error as below:
# deepstream-app -c deepstream_app_config.txt

Using winsys: x11 
ERROR: Deserialize engine failed because file path: /opt/nvidia/deepstream/deepstream-6.0/sources/DeepStream-Yolo-master/model_b1_gpu0_fp32.engine open error
0:00:03.068035757 12438     0x1de33040 WARN                 nvinfer gstnvinfer.cpp:635:gst_nvinfer_logger:&lt;primary_gie&gt; NvDsInferContext[UID 1]: Warning from NvDsInferContextImpl::deserializeEngineAndBackend() &lt;nvdsinfer_context_impl.cpp:1889&gt; [UID = 1]: deserialize engine from file :/opt/nvidia/deepstream/deepstream-6.0/sources/DeepStream-Yolo-master/model_b1_gpu0_fp32.engine failed
0:00:03.068295696 12438     0x1de33040 WARN                 nvinfer gstnvinfer.cpp:635:gst_nvinfer_logger:&lt;primary_gie&gt; NvDsInferContext[UID 1]: Warning from NvDsInferContextImpl::generateBackendContext() &lt;nvdsinfer_context_impl.cpp:1996&gt; [UID = 1]: deserialize backend context from engine from file :/opt/nvidia/deepstream/deepstream-6.0/sources/DeepStream-Yolo-master/model_b1_gpu0_fp32.engine failed, try rebuild
0:00:03.068374641 12438     0x1de33040 INFO                 nvinfer gstnvinfer.cpp:638:gst_nvinfer_logger:&lt;primary_gie&gt; NvDsInferContext[UID 1]: Info from NvDsInferContextImpl::buildModel() &lt;nvdsinfer_context_impl.cpp:1914&gt; [UID = 1]: Trying to create engine from model files

Loading pre-trained weights
deepstream-app: utils.cpp:104: std::vector&lt;float&gt; loadWeights(std::__cxx11::string, const string&amp;): Assertion `count &gt; 0 &amp;&amp; "\nInvalid .wts file."' failed.
Aborted (core dumped)

Then my config file is list：

[property]
gpu-id=0
net-scale-factor=0.0039215697906911373
model-color-format=0
custom-network-config=my_yolov5s.cfg
model-file=my_yolov5s.wts
model-engine-file=model_b1_gpu0_fp32.engine
#int8-calib-file=calib.table
labelfile-path=my_labels.txt
batch-size=1
network-mode=0
num-detected-classes=7
interval=0
gie-unique-id=1
process-mode=1
network-type=0
cluster-mode=2
maintain-aspect-ratio=1
parse-bbox-func-name=NvDsInferParseYolo
custom-lib-path=nvdsinfer_custom_impl_Yolo/libnvdsinfer_custom_impl_Yolo.so
engine-create-func-name=NvDsInferYoloCudaEngineGet

[class-attrs-all]
nms-iou-threshold=0.45
pre-cluster-threshold=0.25
topk=300

I also have other problems，I try another method: I use trt_yolov5.engine with tensort-yolo,my config file is as below
#custom-network-config=my_yolov5s.cfg
#model-file=my_yolov5s.wts
#model-engine-file=model_b1_gpu0_fp32.engine
model-engine-file=trt_yolov5s.engine

then error is WARNING: Num classes mismatch. Configured: 7, detected by network: 0 Segmentation fault (core dumped), at the mean time ,a black screen occurs

Thank you for any kind of help. I will appreciate very much！

**Comment by marcoslucianops** [comment]
Run `sudo apt install --reinstall coreutils` to install `cat` commandm, then convert you model again using `gen_wts_yoloV5.py`.

**Comment by gzcyyds** [comment]
> Run `sudo apt install --reinstall coreutils` to install `cat` command, then convert you model again using the `gen_wts_yoloV5.py`.
sir，very thanks. I'll try it tomorrow and let you know the result. Another question I would like to ask is that I now build pytorch training yolov5 under the local windows system, does it mean that I have to build pytorch environment on jetson XAiver nx for model transformation？

**Comment by marcoslucianops** [comment]
Yes, you need to run it on Linux to do the conversion.

**Comment by stepstep123** [comment]
thanks you great job.

when i customed yolov5 model, i also get problems.

### envs:
deepstream-app version 6.1.0
DeepStreamSDK 6.1.0
Driver Version 515.65
CUDA 11.6

### deepstream-app -c deepstream_app_config.txt

root@3c38dad9654b:/opt/nvidia/deepstream/deepstream-6.1/DeepStream-Yolo# deepstream-app -c deepstream_app_config.txt
WARNING: [TRT]: CUDA lazy loading is not enabled. Enabling it can significantly reduce device memory usage. See `CUDA_MODULE_LOADING` in https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#env-vars
Deserialize yoloLayer plugin: yolo
WARNING: [TRT]: TensorRT was linked against cuDNN 8.6.0 but loaded cuDNN 8.4.0
WARNING: [TRT]: TensorRT was linked against cuDNN 8.6.0 but loaded cuDNN 8.4.0
WARNING: [TRT]: CUDA lazy loading is not enabled. Enabling it can significantly reduce device memory usage. See `CUDA_MODULE_LOADING` in https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#env-vars
0:00:05.144655473 20642 0x7f871c002350 INFO                 nvinfer gstnvinfer.cpp:646:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Info from NvDsInferContextImpl::deserializeEngineAndBackend() <nvdsinfer_context_impl.cpp:1900> [UID = 1]: deserialized trt engine from :/opt/nvidia/deepstream/deepstream-6.1/DeepStream-Yolo/model_b1_gpu0_fp32.engine
INFO: ../nvdsinfer/nvdsinfer_model_builder.cpp:610 [Implicit Engine Info]: layers num: 5
0   INPUT  kFLOAT data            3x640x640
1   OUTPUT kFLOAT num_detections  1
2   OUTPUT kFLOAT detection_boxes 25200x4
3   OUTPUT kFLOAT detection_scores 25200
4   OUTPUT kFLOAT detection_classes 25200

0:00:05.177371613 20642 0x7f871c002350 INFO                 nvinfer gstnvinfer.cpp:646:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Info from NvDsInferContextImpl::generateBackendContext() <nvdsinfer_context_impl.cpp:2003> [UID = 1]: Use deserialized engine model: /opt/nvidia/deepstream/deepstream-6.1/DeepStream-Yolo/model_b1_gpu0_fp32.engine
0:00:05.569673278 20642 0x7f871c002350 INFO                 nvinfer gstnvinfer_impl.cpp:328:notifyLoadModelStatus:<primary_gie> [UID 1]: Load new model:/opt/nvidia/deepstream/deepstream-6.1/DeepStream-Yolo/config_infer_primary_yoloV5.txt sucessfully

Runtime commands:
	h: Print this help
	q: Quit

	p: Pause
	r: Resume

NOTE: To expand a source in the 2D tiled display and view object details, left-click on the source.
      To go back to the tiled display, right-click anywhere on the window.


**PERF:  FPS 0 (Avg)
**PERF:  0.00 (0.00)
** INFO: <bus_callback:194>: Pipeline ready

** INFO: <bus_callback:180>: Pipeline running

**PERF:  268.61 (268.49)
** INFO: <bus_callback:217>: Received EOS. Exiting ...

Quitting
App run successful

### 
[application]
enable-perf-measurement=1
perf-measurement-interval-sec=5

[tiled-display]
enable=1
rows=1
columns=1
width=1280
height=720
gpu-id=0
nvbuf-memory-type=0

[source0]
enable=1
type=3
uri=file:////opt/nvidia/deepstream/deepstream-6.1/yolov5/20211025120613_1080p.mp4
#uri=file:///opt/nvidia/deepstream/deepstream/samples/streams/sample_1080p_h264.mp4
num-sources=1
gpu-id=0
cudadec-memtype=0

[sink0]
enable=1
type=1
sync=0
gpu-id=0
nvbuf-memory-type=0

[osd]
enable=1
gpu-id=0
border-width=5
text-size=15
text-color=1;1;1;1;
text-bg-color=0.3;0.3;0.3;1
font=Serif
show-clock=0
clock-x-offset=800
clock-y-offset=820
clock-text-size=12
clock-color=1;0;0;0
"deepstream_app_config.txt" 64L, 991C                                                                                                                                                     25,1          Top
[application]
enable-perf-measurement=1
perf-measurement-interval-sec=5

[tiled-display]
enable=1
rows=1
columns=1
width=1280
height=720
gpu-id=0
nvbuf-memory-type=0

[source0]
enable=1
type=3
uri=file:////opt/nvidia/deepstream/deepstream-6.1/yolov5/20211025120613_1080p.mp4
#uri=file:///opt/nvidia/deepstream/deepstream/samples/streams/sample_1080p_h264.mp4
num-sources=1
gpu-id=0
cudadec-memtype=0

[application]
enable-perf-measurement=1
perf-measurement-interval-sec=5

[tiled-display]
enable=1
rows=1
columns=1
width=1280
height=720
gpu-id=0
nvbuf-memory-type=0

[source0]
enable=1
type=3
uri=file:////opt/nvidia/deepstream/deepstream-6.1/yolov5/20211025120613_1080p.mp4
#uri=file:///opt/nvidia/deepstream/deepstream/samples/streams/sample_1080p_h264.mp4
num-sources=1
gpu-id=0
cudadec-memtype=0

### config
because i am a ubuntu 18.04 and have no display . so i changed **type=1**.
[sink0]
enable=1
**type=1**
sync=0
gpu-id=0
nvbuf-memory-type=0

root@3c38dad9654b:/opt/nvidia/deepstream/deepstream-6.1/DeepStream-Yolo# cat deepstream_app_config.txt
[application]
enable-perf-measurement=1
perf-measurement-interval-sec=5

[tiled-display]
enable=1
rows=1
columns=1
width=1280
height=720
gpu-id=0
nvbuf-memory-type=0

[source0]
enable=1
type=3
uri=file:////opt/nvidia/deepstream/deepstream-6.1/yolov5/20211025120613_1080p.mp4
#uri=file:///opt/nvidia/deepstream/deepstream/samples/streams/sample_1080p_h264.mp4
num-sources=1
gpu-id=0
cudadec-memtype=0

[sink0]
enable=1
**type=1**
sync=0
gpu-id=0
nvbuf-memory-type=0

[osd]
enable=1
gpu-id=0
border-width=5
text-size=15
text-color=1;1;1;1;
text-bg-color=0.3;0.3;0.3;1
font=Serif
show-clock=0
clock-x-offset=800
clock-y-offset=820
clock-text-size=12
clock-color=1;0;0;0
nvbuf-memory-type=0

[streammux]
gpu-id=0
live-source=0
batch-size=1
batched-push-timeout=40000
width=1920
height=1080
enable-padding=0
nvbuf-memory-type=0

[primary-gie]
enable=1
gpu-id=0
gie-unique-id=1
nvbuf-memory-type=0
config-file=config_infer_primary_yoloV5.txt
#config-file=config_infer_primary.txt

[tests]
file-loop=0


Thank you for any kind of help. I will appreciate very much！

**Comment by stepstep123** [comment]
when i changed the type 1 to 2. i got this

### results:
WARNING: [TRT]: CUDA lazy loading is not enabled. Enabling it can significantly reduce device memory usage. See `CUDA_MODULE_LOADING` in https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#env-vars
Deserialize yoloLayer plugin: yolo
WARNING: [TRT]: TensorRT was linked against cuDNN 8.6.0 but loaded cuDNN 8.4.0
WARNING: [TRT]: TensorRT was linked against cuDNN 8.6.0 but loaded cuDNN 8.4.0
WARNING: [TRT]: CUDA lazy loading is not enabled. Enabling it can significantly reduce device memory usage. See `CUDA_MODULE_LOADING` in https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#env-vars
0:00:05.142236799 20675 0x561beed81e30 INFO                 nvinfer gstnvinfer.cpp:646:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Info from NvDsInferContextImpl::deserializeEngineAndBackend() <nvdsinfer_context_impl.cpp:1900> [UID = 1]: deserialized trt engine from :/opt/nvidia/deepstream/deepstream-6.1/DeepStream-Yolo/model_b1_gpu0_fp32.engine
INFO: ../nvdsinfer/nvdsinfer_model_builder.cpp:610 [Implicit Engine Info]: layers num: 5
0   INPUT  kFLOAT data            3x640x640
1   OUTPUT kFLOAT num_detections  1
2   OUTPUT kFLOAT detection_boxes 25200x4
3   OUTPUT kFLOAT detection_scores 25200
4   OUTPUT kFLOAT detection_classes 25200

0:00:05.202643303 20675 0x561beed81e30 INFO                 nvinfer gstnvinfer.cpp:646:gst_nvinfer_logger:<primary_gie> NvDsInferContext[UID 1]: Info from NvDsInferContextImpl::generateBackendContext() <nvdsinfer_context_impl.cpp:2003> [UID = 1]: Use deserialized engine model: /opt/nvidia/deepstream/deepstream-6.1/DeepStream-Yolo/model_b1_gpu0_fp32.engine
0:00:05.592831762 20675 0x561beed81e30 INFO                 nvinfer gstnvinfer_impl.cpp:328:notifyLoadModelStatus:<primary_gie> [UID 1]: Load new model:/opt/nvidia/deepstream/deepstream-6.1/DeepStream-Yolo/config_infer_primary_yoloV5.txt sucessfully
** ERROR: <main:733>: Could not open X Display
Quitting
App run failed

thanks again.

**Comment by marcoslucianops** [comment]
The first case isn't a problem. It got EOF and stopped the file playing. To process it slowly, you need to set `[sink] sync=1`.

The second case is because you aren't using screen. The `[sink] type=2` requires screen connected.

**Comment by stepstep123** [comment]
> 

thanks your reply and help.

so if i want to test the FPS when i have no screen. i just set [sink] sync=0 and type=2. 
**PERF: 268.61 (268.49) is the current FPS of model?

**Comment by stepstep123** [comment]
> 

i guess i finished the deepstream-app -c deepstream_app_config.txt.
but how can i save a local file as the output? 
do you have any suggestions or code demo?

thanks your help.

**Comment by marcoslucianops** [comment]
> **PERF: 268.61 (268.49) is the current FPS of model?

Yes, it's.

> how can i save a local file as the output?

For `deepstream-app`, you can add these lines in the `deepstream_app_config.txt` file

```
[sink1]
enable=1
type=3
container=1
codec=1
enc-type=0
sync=0
bitrate=4000000
profile=0
output-file=output.mp4
source-id=0
```
